src/test/kotlin/org/rust/cargo/toolchain/CargoProjectGenerationTest.kt

```diff
@@ -0,0 +1,124 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.cargo.toolchain
+
+import com.intellij.openapi.vfs.VirtualFile
+import com.intellij.testFramework.fixtures.impl.TempDirTestFixtureImpl
+import org.rust.Entry
+import org.rust.FileTree
+import org.rust.cargo.RsWithToolchainTestBase
+import org.rust.cargo.project.settings.toolchain
+import org.rust.cargo.toolchain.tools.Cargo
+import org.rust.cargo.toolchain.tools.cargo
+import org.rust.fileTree
+import org.rust.ide.newProject.RsCustomTemplate
+import org.rust.ide.newProject.RsGenericTemplate
+import org.rust.ide.newProject.RsProjectTemplate
+import org.rust.ide.newProject.makeProject
+import org.rust.openapiext.modules
+import org.rust.openapiext.saveAllDocuments
+
+class CargoProjectGenerationTest : RsWithToolchainTestBase() {
+    private val tempDirFixture: TempDirTestFixtureImpl = TempDirTestFixtureImpl()
+
+    private val cargo: Cargo
+        get() = project.toolchain?.cargo()!!
+
+    override fun setUp() {
+        super.setUp()
+        tempDirFixture.setUp()
+    }
+
+    override fun tearDown() {
+        tempDirFixture.tearDown()
+        super.tearDown()
+    }
+
+    fun `test binary project generation`() = fileTree {
+        dir("src") {
+            file("main.rs", "")
+        }
+        file("Cargo.toml", "")
+    }.checkTemplate(RsGenericTemplate.CargoBinaryTemplate)
+
+    fun `test library project generation`() = fileTree {
+        dir("src") {
+            file("lib.rs", "")
+        }
+        file("Cargo.toml", "")
+    }.checkTemplate(RsGenericTemplate.CargoLibraryTemplate)
+
+    fun `test proc macro project generation`() = fileTree {
+        dir("src") {
+            file("lib.rs", "")
+        }
+        file("Cargo.toml", "")
+    }.checkTemplate(RsCustomTemplate.ProcMacroTemplate)
+
+    fun `test wasm-pack project generation`() = fileTree {
+        dir("src") {
+            file("lib.rs", "")
+            file("utils.rs", "")
+        }
+        file("Cargo.toml", "")
+    }.checkTemplate(RsCustomTemplate.WasmPackTemplate)
+
+    fun `test cargo init with git`() {
+        cargo.init(project, testRootDisposable, tempDirFixture.getFile("")!!, "foo", false, "git")
+        assertTrue(listOf(".gitignore", ".git").map { tempDirFixture.getFile(it) }.all { it != null })
+    }
+
+    fun `test cargo init without git`() {
+        cargo.init(project, testRootDisposable, tempDirFixture.getFile("")!!, "foo", false, "none")
+        assertTrue(listOf(".gitignore", ".git").map { tempDirFixture.getFile(it) }.all { it == null })
+    }
+
+    private fun FileTree.checkTemplate(template: RsProjectTemplate) {
```

Comment:
nit: I think with the following signature, we can produce a little but nicer code:
```kotlin
private fun checkTemplate(template: RsProjectTemplate, builder: FileTreeBuilder.() -> Unit) {
    ...
    val fileTree = fileTree(builder)
    ...
}
```
Example of possible usage:
```kotlin
fun `test proc macro project generation`() = checkTemplate(RsCustomTemplate.ProcMacroTemplate) {
    dir("src") {
        file("lib.rs", "")
    }
    file("Cargo.toml", "")
}
```